sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+    }
+    
+    // Initialize each TOF bin's 2D histogram
+    for (auto& tof_image : tof_images) {
+        tof_image.resize(dim_y, std::vector<unsigned int>(dim_x, 0));
+    }
+
+    // Process neutrons from all batches
+    size_t total_entries = 0;
+    size_t binned_entries = 0;
+
+    for (size_t batch_index = 0; batch_index < batches.size(); ++batch_index) {
+        const auto& batch = batches[batch_index];
+        spdlog::debug("Processing batch {}", batch_index);
+
+        std::vector<const IPositionTOF*> entries;
+        if (mode == "hit") {
+            entries.reserve(batch.hits.size());
+            for (const auto& hit : batch.hits) {
+                entries.push_back(static_cast<const IPositionTOF*>(&hit));
+            }
+        } else {
+            entries.reserve(batch.neutrons.size());
+            for (const auto& neutron : batch.neutrons) {
+                entries.push_back(static_cast<const IPositionTOF*>(&neutron));
+            }
+        }
+        
+        if (entries.empty()) {
+            spdlog::debug("Batch {} is empty", batch_index);
+            continue;
+        }
+
+        for (const auto& entry : entries) {
+            total_entries++;
+            double tof_ns = entry->iGetTOF_ns();
+
+            // Find the correct TOF bin
+            // NOTE: tof_bin_edges are in sec, and tof_ns are in nano secs
+            spdlog::debug("tof_ns: {}, tof_ns/1e9: {}", tof_ns, tof_ns/1e9);
+            auto it = std::lower_bound(tof_bin_edges.begin(), tof_bin_edges.end(), tof_ns/1e9);
+            if (it != tof_bin_edges.begin()) {
+                size_t bin_index = std::distance(tof_bin_edges.begin(), it) - 1;
+                
+                // Calculate the x and y indices in the 2D histogram
+                int x = static_cast<int>(entry->iGetX() * super_resolution);
+                int y = static_cast<int>(entry->iGetY() * super_resolution);
+                
+                // Ensure x and y are within bounds
+                if (x >= 0 && x < dim_x && y >= 0 && y < dim_y) {
+                    // Increment the count in the appropriate bin and position
+                    tof_images[bin_index][y][x]++;
+                    binned_entries++;
+                }
+            }
+        }
+    }
+
+    auto end = std::chrono::high_resolution_clock::now();
+    auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+    spdlog::info("TOF image creation time: {} s", elapsed / 1e6);
+    spdlog::info("Total entries: {}, Binned entries: {}", total_entries, binned_entries);
+
+    return tof_images;
+}
+
+/**
+ * @brief Timed save TOF imaging to TIFF.
+ * 
+ * @param[in] out_tof_imaging
+ * @param[in] batches
+ * @param[in] tof_bin_edges
+ * @param[in] tof_filename_base
+ */
+void timedSaveTOFImagingToTIFF(
+    const std::string& out_tof_imaging,
+    const std::vector<std::vector<std::vector<unsigned int>>>& tof_images,
```

Comment:
This wants a typedef to capture the concept with a name

Reply:
use macro as suggested